app/src/main/java/com/simplecity/amp_library/playback/MusicService.java

```diff
@@ -1581,28 +1674,100 @@ public void moveQueueItem(int from, int to) {
         }
     }
 
-    public void makeShuffleList() {
+    public void makeTrackShuffleList() {
         synchronized (this) {
 
             if (playlist == null || playlist.isEmpty()) {
                 return;
             }
 
-            shuffleList = new ArrayList<>(playlist);
+            trackShuffleList = new ArrayList<>(playlist);
             Song currentSong = null;
-            if (playPos >= 0 && playPos < shuffleList.size()) {
-                currentSong = shuffleList.remove(playPos);
+            if (playPos >= 0 && playPos < trackShuffleList.size()) {
+                currentSong = trackShuffleList.remove(playPos);
             }
 
-            Collections.shuffle(shuffleList);
+            Collections.shuffle(trackShuffleList);
 
             if (currentSong != null) {
-                shuffleList.add(0, currentSong);
+                trackShuffleList.add(0, currentSong);
             }
             playPos = 0;
         }
     }
 
+    public void makeAlbumShuffleList() {
+        synchronized (this) {
+            if (albumShuffleList.size() > 0) {
+                albumShuffleList = new ArrayList<>();
+            }
+
+            if (playlist == null || playlist.isEmpty()) {
+                return;
+            }
+
+            Map<Album, List<Song>> trackShuffleListAlbumMap = new HashMap<>();
+
+            for (Song song : trackShuffleList) {
+                Album containingAlbum = song.getAlbum();
+                if (!trackShuffleListAlbumMap.containsKey(containingAlbum)) {
+                    trackShuffleListAlbumMap.put(containingAlbum, new ArrayList<>());
+                }
+                trackShuffleListAlbumMap.get(containingAlbum).add(song);
+            }
+
+            Song currentSong = null;
+            if (playPos >= 0 && playPos < playlist.size()) {
+                currentSong = trackShuffleList.get(playPos);
+            }
+
+            if (currentSong != null) {
+                trackShuffleListAlbumMap.get(currentSong.getAlbum()).remove(currentSong);
+                albumShuffleList.add(currentSong);
+                albumShuffleList.addAll(trackShuffleListAlbumMap.get(currentSong.getAlbum()));
+                trackShuffleListAlbumMap.remove(currentSong.getAlbum());
+            }
+
+            List<Album> trackShuffleListAlbums = new ArrayList<>(trackShuffleListAlbumMap.keySet());
+            Collections.shuffle(trackShuffleListAlbums);
+
+            for (Album album : trackShuffleListAlbums) {
+                albumShuffleList.addAll(trackShuffleListAlbumMap.get(album));
+            }
+
+            playPos = 0;
+        }
+    }
+
+    private static List<Song> albumShuffleify(List<Song> songs) {
```

Comment:
This method could use some comments on what it does, as it's not clear from the method name alone.

'Takes a list of songs, groups them by album, then shuffles the albums - returning the resultant 'album-shuffled' list'

Perhaps the method could also be renamed to `albumShuffleSongs` (though it's not a huge improvement!)